pennylane/transforms/decompositions/single_qubit_unitary.py

```diff
@@ -81,27 +71,20 @@ def zyz_decomposition(U, wire):
     """
     U = _convert_to_su2(U)
 
-    # Check if the matrix is diagonal; only need to check one corner.
-    # If it is diagonal, we don't need a full Rot, just return an RZ.
-    if math.allclose(U[0, 1], [0.0]):
-        omega = 2 * math.angle(U[1, 1])
-        return [qml.RZ(omega, wires=wire)]
-
-    # If the top left element is 0, can only use the off-diagonal elements. We
-    # have to be very careful with the math here to ensure things that get
-    # multiplied together are of the correct type in the different interfaces.
-    if math.allclose(U[0, 0], [0.0]):
-        phi = 0.0
-        theta = -np.pi
-        omega = 1j * math.log(U[0, 1] / U[1, 0]) - np.pi
-    else:
-        # If not diagonal, compute the angle of the RY
-        cos2_theta_over_2 = math.abs(U[0, 0] * U[1, 1])
-        theta = 2 * math.arccos(math.sqrt(cos2_theta_over_2))
-
-        el_division = U[0, 0] / U[1, 0]
-        tan_part = math.cast_like(math.tan(theta / 2), el_division)
-        omega = 1j * math.log(tan_part * el_division)
-        phi = -omega - math.cast_like(2 * math.angle(U[0, 0]), omega)
-
-    return [qml.Rot(math.real(phi), math.real(theta), math.real(omega), wires=wire)]
+    # Derive theta from the off-diagonal element. We very very carefully rescale
+    # the value to make sure it's between -1 and 1 by subtracting a small number
+    # in the cases where we have a positive-valued element.
+    element = math.abs(U[0, 1])
+    theta = 2 * math.arcsin(element - math.sign(element) * 1e-16)
+
+    # Compute phi and omega from the angles of the top row; use atan2 to keep
+    # the angle within -np.pi and np.pi, and add very small values to avoid the
+    # undefined case of 0/0. We add a smaller value to the imaginary part than
+    # the real part because it is imag / real in the definition of atan2.
+    angle_U00 = math.arctan2(math.imag(U[0, 0]) + 1e-128, math.real(U[0, 0]) + 1e-64)
+    angle_U10 = math.arctan2(math.imag(U[1, 0]) + 1e-128, math.real(U[1, 0]) + 1e-64)
```

Comment:
`1e-128` enforces the use of double precision I suppose? Or is it beyond double precision?

Reply:
Oh I just had a thought - these very small values that are being added/subtracted, is this the cause of the reduction in precision at the end?